GraalVM support with testing

maschnetwork opened the pull request:
*Issue # and related PRs*

#464 
#311

And larger issue on the AWS SDK repo: 

https://github.com/aws/aws-sdk-java-v2/issues/2948

**Description of changes:**

Added configuration files (in META-INF/native-image) for latest version of CRT client to support GraalVM native images. Added Maven Profile (graalvm-native) for running unit tests as a native image which should verify that all necessary configuration is present. This would allow us to identify if any configuration is missing in the future. I did not include it into any build pipeline and will rely on the maintainers guidance on how to implement this check properly. The command to run the unit tests with the profile is:

```bash
mvn clean test -Pgraalvm-native 
```

The native testing method requires Junit 5 which is why the profile is leveraging the [JUnitVintage Engine](https://mvnrepository.com/artifact/org.junit.vintage/junit-vintage-engine). Also I had to add the src/main/resources folder to be recognized in the pom - otherwise the configuration files were ignored during test runs. Lastly, I added a check `skipIfNativeImage()` to skip JVM related tests as they would not apply to a native binary.

**How were the configuration files created?**

The files were generated by attaching the [native image tracing agent](https://www.graalvm.org/latest/reference-manual/native-image/metadata/AutomaticMetadataCollection/) to the maven-surefire-plugin via the argline command. This will generate the configuration based on the unit tests that are executed. I also had to set forkCount to 1 otherwise the agent was not started:

```xml
<!-- unit test configuration -->
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-surefire-plugin</artifactId>
    <version>3.2.3</version>
    <configuration>
          <!-- Other configuration ommited for readability -->
          <argLine>-agentlib:native-image-agent=config-merge-dir=native-configuration -Daws.crt.memory.tracing=2 -Xcheck:jni</argLine>
         <forkCount>1</forkCount>
      </configuration>
</plugin>
```
Since the generated configuration also includes test classes that are not needed for the final artifact I haven't automated this task and manually diffed and merged them into the final META-INF/native-image folder.

A fully working version (with the configuration being manually added) can be found here: https://github.com/maschnetwork/aws-lambda-graalvm-crt - I also built a SNAPSHOT jar locally and tested it with both Sync and Async client for DynamoDB.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

deki reviewed (approved):
Nice work!

maschnetwork commented on src/main/resources/META-INF/native-image/software.amazon.awssdk/crt/aws-crt/reflect-config.json:
```diff
@@ -0,0 +1,60 @@
+    "methods":[{"name":"getSuppressed","parameterTypes":[] }]
+  },
+  {
+    "name":"java.util.concurrent.ForkJoinTask",
```
Those are files that have been generated by running the Unit-Tests and some of them seem to be related to test cases only. Those should be not needed as the native-image tests require a normal test run in advance so that those files are automatically created. I cleaned up the obvious ones and will push an update.

maschnetwork commented on pom.xml:
```diff
@@ -216,6 +216,60 @@
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>org.junit.vintage</groupId>
+                    <artifactId>junit-vintage-engine</artifactId>
+                    <version>5.10.1</version>
```
Agree. Moved Vintage Engine out of the profile.

maschnetwork commented on src/main/resources/META-INF/native-image/software.amazon.awssdk/crt/aws-crt/native-image.properties:
```diff
@@ -0,0 +1 @@
+Args=--initialize-at-run-time=io.netty.util.internal.logging.Log4JLogger -H:ClassInitialization=org.slf4j:build_time
```
Was copied from the previous PR (https://github.com/awslabs/aws-crt-java/pull/464/files) - could not reproduce the warning and will therefore remove.

klopfdreh commented:
I recently tested the jni-config.json on x86_64 / linux and an upload to s3 with the AWS CRT dependency (`S3AsyncClient.crtBuilder()`) like 
`native-image ... -H:JNIConfigurationFiles=META-INF/native-image/spring-boot-jdk17-x86-native/jni-config.json`
and it seems to work!

I don't know if the JNI configuration have been changed again as there are many "Waiting for status to be reported" and the branch is "out-of-date"

![image](https://github.com/awslabs/aws-crt-java/assets/980773/4238ac18-beef-475a-ba3b-720a7bb424f5)
![image](https://github.com/awslabs/aws-crt-java/assets/980773/59649df5-3d0e-4752-ba67-14b1d7f09dcd)

nice work anyway @maschnetwork 👍

klopfdreh commented:
Hey @maschnetwork maybe it would be a good idea to introduce the native support to the code as in `Spring Cloud AWS` - see: https://github.com/awspring/spring-cloud-aws/pull/856/files

maschnetwork commented:
Hi @klopfdreh, thanks for reaching out. I will update the PR shortly and will add missing configuration files (if there are any). Thanks for testing and verifying that it is working for you.

klopfdreh commented:
@maschnetwork - thanks a lot! Just one hint: When I tried it with the newest version of AWS SDK V2 for Java there is a runtime issue that a method is missing in the S3AsyncClient when using crtBuilder.

maschnetwork commented:
@klopfdreh I updated the branch with the new configuration file. There was one configuration missing _software.amazon.awssdk.crt.http.HttpProxyOptions_ which has been added. Can you please confirm your setup is working with the new configuration?

@graebm I now also added a new maven profile _generate-graalvm-files_ that runs the unit tests with the tracing agent to create the new configuration file. Verifying and adding the missing configuration is now relatively straight forward. I will outline the example on how i created the new version below:

1. Initially I ran _mvn clean test -Pgraalvm-native_ which failed as expected as there were configuration files missing:

![Bildschirmfoto 2024-05-07 um 07 14 48](https://github.com/awslabs/aws-crt-java/assets/36627945/83109e47-a09a-46ba-b1ac-213cd7b06683)

2. It correctly pointed me to java_class_ids.c which has been changed recently:
![image](https://github.com/awslabs/aws-crt-java/assets/36627945/cb324237-8fd4-4129-b371-8df4e3891dbb)

3. I then ran _mvn test -Pgenerate-graalvm-files_ to generate the new configuration files and diffed them with the existing ones and I identified correctly that the following configuration was missing:

`{
  "name":"software.amazon.awssdk.crt.http.HttpProxyOptions$HttpProxyAuthorizationType",
  "methods":[{"name":"getValue","parameterTypes":[] }]
}`

4. I added this to the src/resources/META-INF/../jni-config.json and reran _mvn clean test -Pgraalvm-native_ which then resulted in a successful build.

klopfdreh commented:
@maschnetwork - I am going to test it with the newest AWS SDK V2 and CRT with those configs and give feedback later this day. 👍

klopfdreh commented:
@maschnetwork - I used the following versions:

```xml
        <awssdk-v2.version>2.25.46</awssdk-v2.version>
        <aws-crt.version>0.29.18</aws-crt.version>
```

1. native-build was a success ✔️ 
2. I received a runtime exception when I tried to get resources from S3 ❌ 
```plain
java.lang.NoSuchMethodError: software.amazon.awssdk.identity.spi.Identity.providerName()
at software.amazon.awssdk.core.internal.http.pipeline.stages.ApplyUserAgentStage.providerNameFromIdentity(ApplyUserAgentStage.java:174) ~[na:na]
at software.amazon.awssdk.core.internal.http.pipeline.stages.ApplyUserAgentStage.identityProviderName(ApplyUserAgentStage.java:168) ~[na:na]
at software.amazon.awssdk.core.internal.http.pipeline.stages.ApplyUserAgentStage.getUserAgent(ApplyUserAgentStage.java:147) ~[na:na]
at software.amazon.awssdk.core.internal.http.pipeline.stages.ApplyUserAgentStage.execute(ApplyUserAgentStage.java:135) ~[na:na]
at software.amazon.awssdk.core.internal.http.pipeline.stages.ApplyUserAgentStage.execute(ApplyUserAgentStage.java:52) ~[na:na]
```

I guess there is a hint missing in a reflect-config.json in the core module at AWS SDK V2.

maschnetwork commented:
@klopfdreh Yes this is not related to the CRT client configuration but to the AWS SDK v2: https://github.com/aws/aws-sdk-java-v2 - I see that providerName has been recently added there. Can you please open an issue there with your example code and logs? I will take care of communicating it internally and can also take a look into providing a fix.

klopfdreh commented:
> @klopfdreh Yes this is not related to the CRT client configuration but to the AWS SDK v2: https://github.com/aws/aws-sdk-java-v2 - I see that providerName has been recently added there. Can you please open an issue there with your example code and logs? I will take care of communicating it internally and can also take a look into providing a fix.

Hey @maschnetwork - I double checked my dependencies and found `identity-spi` version was not set correctly. I just fixed that and can confirm that all is running correctly! 🎉

I used the following code to upload a 145MB file to S3 and all is working within a native image. 👍 

```java
try (S3AsyncClient s3AsyncClient = S3AsyncClient.crtBuilder()
        .credentialsProvider(awsCredentialsProvider)
        .region(region)
        .endpointOverride(endpoint)
        .targetThroughputInGbps(20.0)
        .minimumPartSizeInBytes(20000000L)
        .maxConcurrency(8)
        .build();
     S3TransferManager s3TransferManager = S3TransferManager
             .builder()
             .s3Client(s3AsyncClient)
             .build()) {
 // ...
 }
```

klopfdreh reviewed (approved):
Test native image s3 upload.

graebm commented:
I chatted with @maschnetwork.

Before this is merged, **ALL** reflection calls need to be moved into functions that are called on startup. Most reflection calls already happen this way, but I want to be sure we're not missing anything. This way, we'll detect missing entries in `META-INF/native-image/...` **immediately**, without requiring 100% branch coverage in our tests.

If we do this, maybe just remove any automation for generating `META-INF/native-image`? Since, from hereon, it will just be single line additions? Less code is best code.

We'll also need CI for GraalVM, so we don't break this moving forward. I suppose this can be added in a subsequent PR.

klopfdreh commented:
Thanks a lot! When ever I can test something regarding the native build, please ping me at this PR. 👍

TingDaoK commented:
Hey @maschnetwork , I tried to run `mvn clean test -Pgraalvm-native`, and they all failed with different errors for each version of java. You can find the details [here](https://github.com/awslabs/aws-crt-java/actions/runs/9165954012/job/25200407977?pr=786).

For java-version 22, I guess there is a new package was added and the config file needs to be updated.

For the others, I assume it's s3 tests was not running when you generate the config file. However, when I tried to run `mvn test -Pgenerate-graalvm-files`, I got the following errors from the S3 tests:
```
[INFO] Running software.amazon.awssdk.crt.test.S3ClientTest
/bin/sh: line 1: 23624 Abort trap: 6           '/Library/Java/JavaVirtualMachines/graalvm-jdk-17.0.11+7.1/Contents/Home/bin/java' '-agentlib:native-image-agent=config-output-dir=graalvm-configuration' '-Daws.crt.memory.tracing=2' '-Xcheck:jni' '-jar' '/Users/dengket/project/crts/aws-crt-java/target/surefire/surefirebooter-20240520155148744_88.jar' '/Users/dengket/project/crts/aws-crt-java/target/surefire' '2024-05-20T15-51-48_221-jvmRun1' 'surefire-20240520155148744_86tmp' 'surefire_28-20240520155148744_87tmp'
Picked up JAVA_TOOL_OPTIONS: -Dlog4j2.formatMsgNoLookups=true
native-image-agent: Error: Output directory 'graalvm-configuration' is locked by process 23624, which means another agent instance is already writing to this directory. Only one agent instance can safely write to a specific target directory at the same time. Unless file '.lock' is a leftover from an earlier process that terminated abruptly, it is unsafe to delete it. For running multiple processes with agents at the same time to create a single configuration, read AutomaticMetadataCollection.md or https://www.graalvm.org/dev/reference-manual/native-image/metadata/AutomaticMetadataCollection/ on how to use the native-image-configure tool.
[ERROR] Error occurred during initialization of VM
Picked up JAVA_TOOL_OPTIONS: -Dlog4j2.formatMsgNoLookups=true
native-image-agent: Error: Output directory 'graalvm-configuration' is locked by process 23624, which means another agent instance is already writing to this directory. Only one agent instance can safely write to a specific target directory at the same time. Unless file '.lock' is a leftover from an earlier process that terminated abruptly, it is unsafe to delete it. For running multiple processes with agents at the same time to create a single configuration, read AutomaticMetadataCollection.md or https://www.graalvm.org/dev/reference-manual/native-image/metadata/AutomaticMetadataCollection/ on how to use the native-image-configure tool.[ERROR] Error occurred during initialization of VM

Picked up JAVA_TOOL_OPTIONS: -Dlog4j2.formatMsgNoLookups=true
[ERROR] Error occurred during initialization of VMnative-image-agent: Error: Output directory 'graalvm-configuration' is locked by process 23624, which means another agent instance is already writing to this directory. Only one agent instance can safely write to a specific target directory at the same time. Unless file '.lock' is a leftover from an earlier process that terminated abruptly, it is unsafe to delete it. For running multiple processes with agents at the same time to create a single configuration, read AutomaticMetadataCollection.md or https://www.graalvm.org/dev/reference-manual/native-image/metadata/AutomaticMetadataCollection/ on how to use the native-image-configure tool.
```

I guess we are creating a separate thread group for S3, and the tool cannot handle multi-thread properly? 

I'll keep digging into it, just sharing my experiments.

TingDaoK commented:
> Lastly, I added a check skipIfNativeImage() to skip JVM related tests as they would not apply to a native binary.

Can you elaborate a bit more about why they are not apply to the native binary?

TingDaoK commented:
> @TingDaoK Thanks for setting this up. I can't reproduce the failed S3 Tests locally because the tests are not being executed for some reason (Says ABORTED for most of the S3ClientTest). I tried to set AWS credentials before running tests but then even normal mvn tests fail at testCreateDestroyProfile_ValidCreds. I can't find anything in the documentation on what needs to be setup to run the full test suite locally - do you have any hints?
> 
> As for the other issue with `mvn test -Pgenerate-graalvm-files` it seems like another process locked the generation folder. Simply deleting it and running it again should solve it.



> @TingDaoK Thanks for setting this up. I can't reproduce the failed S3 Tests locally because the tests are not being executed for some reason (Says ABORTED for most of the S3ClientTest). I tried to set AWS credentials before running tests but then even normal mvn tests fail at testCreateDestroyProfile_ValidCreds. I can't find anything in the documentation on what needs to be setup to run the full test suite locally - do you have any hints?

Yeah, we plan to add a helper, but we haven't got chance to do so. I'll add the helper tomorrow, so that you can run the tests with your account.


> As for the other issue with `mvn test -Pgenerate-graalvm-files` it seems like another process locked the generation folder. Simply deleting it and running it again should solve it.

I don't think I have any other process running in the background. I guess we create two thread group for s3 client?

TingDaoK commented on pom.xml:
```diff
@@ -214,6 +214,46 @@
+                                </goals>
+                                <phase>test</phase>
+                            </execution>
+                            <execution>
+                                <id>build-native</id>
+                                <goals>
+                                    <goal>build</goal>
+                                </goals>
+                                <phase>package</phase>
+                            </execution>
```
Just want to get your opinion, do we need the build phase? Should it be the application using our lib to build the native executable?

From my understanding, we will not package our lib as a native lib for other package to use, right??

maschnetwork replied:
Correct, for the current setup we don't need it. Let's remove it together with the tracing agent generation if we agree on the json vs. class format for the hints.

maschnetwork commented:
Hi @graebm and @TingDaoK, can you please take a look into the latest commit [094af5e](https://github.com/awslabs/aws-crt-java/pull/749/commits/094af5edad09b516424528480fc4179a2a557a75)? This is an example of leveraring the Feature implementation and profived the JNI configs via a Java class. Wdyt?

maschnetwork commented:
For transparency : After investigation together with @TingDaoK the CI builds are now working: https://github.com/awslabs/aws-crt-java/pull/786/checks.

TingDaoK commented:
> Hi @graebm and @TingDaoK, can you please take a look into the latest commit [094af5e](https://github.com/awslabs/aws-crt-java/pull/749/commits/094af5edad09b516424528480fc4179a2a557a75)? This is an example of leveraging the Feature implementation to provide the JNI configs via a Java class. Wdyt?

With the the Feature implementation, can we use Java reflection to iterator through the classes and methods from`software.amazon.awssdk.crt`, and register them all?

But, looks like the CI fails for Java 17 `Specified module 'org.graalvm.nativeimage' is unknown`.
Also, for Java 8, looks like graalvm SDK wants Java 11+, but we need to support Java 8.
https://github.com/awslabs/aws-crt-java/actions/runs/9354653868/job/25747987564?pr=786
We need to solve those issues before we keep investigating the the `Feature` approach. 

I am fine with the current .json config approach, it's just an extra step to add the binding. And if we forget to add the config, it should be caught by the CI.

maschnetwork commented:
@TingDaoK based on your feedback I reverted the last commit and we stick with the config.json for now - also removed the build execution from the POM. If CI is passing - is there anything left from your side?

TingDaoK commented:
@maschnetwork I just took over the branch to keep it simple. Mostly, I removed the unnecessary reflection from `reflect-config.json` based on my tests. And removed `native-image.properties` as we are not using lf4j. And several trivial fix for the tests we have. Let me know if you want to keep this PR instead of https://github.com/awslabs/aws-crt-java/pull/791

maschnetwork commented:
@TingDaoK Great - even better if it works with less config for the reflection. On this PR I see that you have added several FATAL_ASSERT to the c file. I don't see them anymore on your PR - does it make sense to add them there? [src/native/java_class_ids.c](https://github.com/awslabs/aws-crt-java/pull/749/files#diff-d2c36c154758ebb9053ac6119ddba0ec52bdac375c6beac2a4a0d8ba5d966e43)

TingDaoK commented:
> @TingDaoK Great - even better if it works with less config for the reflection. On this PR I see that you have added several FATAL_ASSERT to the c file. I don't see them anymore on your PR - does it make sense to add them there? [src/native/java_class_ids.c](https://github.com/awslabs/aws-crt-java/pull/749/files#diff-d2c36c154758ebb9053ac6119ddba0ec52bdac375c6beac2a4a0d8ba5d966e43)

yeah, I have already merged that change into main as a separate PR, [here](https://github.com/awslabs/aws-crt-java/pull/789), since I wanted the fix to be merged quicker, it should affect the regular JVM as well.

maschnetwork commented:
@TingDaoK Great - anything else missing then?

TingDaoK commented:
@maschnetwork nothing blocking, thanks for your contributing!

merged with https://github.com/awslabs/aws-crt-java/pull/791

klopfdreh commented:
@TingDaoK and @maschnetwork - thanks a lot for all the work - maybe it would be good to add a hint in the README.md on how to use related json files during a native build. WDYT?

maschnetwork commented:
@TingDaoK Great  - thank you for all your effort on the CI/Testing side and the final cleanup

> how to use the related json files during a native build

@klopfdreh Can you elaborate on this? As the config files now ship with CRT itself there should be no action or manual config changes needed on the client/user side of the library.

klopfdreh commented:
@maschnetwork - sure - I assumed like mentioned here https://aws.amazon.com/de/blogs/developer/graalvm-native-image-support-in-the-aws-sdk-for-java-2-x/ we have to add the following arguments to the `native-image` command for a native build:

```shell
native-image \
-H:JNIConfigurationFiles=META-INF/native-image/software.amazon.awssdk/crt/aws-crt/jni-config.json  \
-H:ReflectionConfigurationResources=META-INF/native-image/software.amazon.awssdk/crt/aws-crt/reflect-config.json  \
-H:ResourceConfigurationResources=META-INF/native-image/software.amazon.awssdk/crt/aws-crt/resource-config.json \
```

Those information I wanted to put in the README.md to ensure perform the native build correctly without any issues.

maschnetwork commented:
@klopfdreh If the configuration files are placed (by the library owner) under the default GraalVM folder e.g. _resources/META_INF/native-image/the_library_package_ - then no additional configuration should be needed on your side.  They should be automatically picked up.

klopfdreh commented:
@maschnetwork - all right 👍

maschnetwork commented:
🎉 Merged with #791. SDK Related PR to bump the version: https://github.com/aws/aws-sdk-java-v2/pull/5290.

Thanks for all the contributions here!
